### ISSUE TITLE
Recycle trap

### ISSUE THREAD

**Issue by clinuxrulz** [h908714124/zerobuilder#3]
I got caught out using "recycle = true". I was filling out fields using the generated step builder, but the fields were set by evaluation of methods and these methods deep down internally used the same step builder on the same thread making a mess of the field values.

I was wondering if its possible to use reference counting to make it behave like "recycle = false" when the reference count is more than one, even if "recycle = true". Just to help people avoid the same trap.

Or even memory pools for handling multiple instances of the step builder in the same scope. (A scope being defined from the allocation of the builder instance to the last field being set)

**Comment by jbgi** [comment]
I think the easiest way is to `ThreadLocal#remove` the builder instance just after the `ThreadLocal#get`, before returning it to user code.
And then, at the last step. just before returning the built instance, put it in back via `ThreadLocal#set(this)`.

**Comment by h908714124** [comment]
Hi @clinuxrulz, could you provide an example? Cheers

**Comment by jbgi** [comment]
@h908714124, here is an example:
```java
public static void main(String[] args) {

    RequestBuilders.PutBuilder.Body builder1 = RequestBuilders.putBuilder()
        .path("/path/1");

    RequestBuilders.putBuilder()
        .path("/path/2");

    Request request = builder1.body("body");

    // print "/path/2" but should be "/path/1":
    System.out.println(Requests.getPath(request)); 
  }
```

**Comment by jbgi** [comment]
@h908714124, and here is what my above suggestion would mean in term of generated code:
```java
public final class RequestBuilders {

  private static final ThreadLocal<PutBuilderImpl> PutBuilderImpl = new ThreadLocal<PutBuilderImpl>() {
    @Override
    protected PutBuilderImpl initialValue() {
      return new PutBuilderImpl();
    }
  };

  private RequestBuilders() {
    throw new UnsupportedOperationException("no instances");
  }

  public static PutBuilder.Path putBuilder() {
    PutBuilderImpl putBuilderImpl = PutBuilderImpl.get();
    PutBuilderImpl.remove();
    return putBuilderImpl;
  }

  private static final class PutBuilderImpl implements PutBuilder.Path, PutBuilder.Body {
    private String path;

    PutBuilderImpl() {
    }

    @Override
    public PutBuilder.Body path(String path) {
      this.path = path;
      return this;
    }

    @Override
    public Request body(String body) {
      String path = this.path;
      this.path = null;
      PutBuilderImpl.set(this);
      return Request.put(path, body);
    }
  }

  public static final class PutBuilder {
    private PutBuilder() {
      throw new UnsupportedOperationException("no instances");
    }

    public interface Path {
      Body path(String path);
    }

    public interface Body {
      Request body(String body);
    }
  }

}
```

**Comment by h908714124** [comment]
@jbgi imho `PutBuilderImpl.remove()` would cost some performance. If you're willing to trade performance for safety, then why would you use `recycle` in the first place?

@clinuxrulz I think you can still use the `recycle` option, as long as you follow this rule regarding the generated types:

* Never declare a variable, field or method parameter of type `XBuilders.YBuilder.Foo` or `XBuilders.YUpdater`.

For example, in JB's first comment, the variable `builder1` is breaking the rule.

**Comment by jbgi** [comment]
@h908714124 Here is an example that does not break the rule:
```java
public static void main(String[] args) {
    Request request = RequestBuilders.putBuilder()
        .path("/external/path")
        .body(Requests.getBody(RequestBuilders.putBuilder().path("/internal/path").body("body")).orElse(""));

    // print "/internal/path" but should be "/external/path":
    System.out.println(Requests.getPath(request));
  }
```
For me this inability to safely "compose" usage of recycled builders is a show-stopper. If indeed `remove()` (or another ref-counting technique) incurs a greater cost that creating new builder instances (would need to benchmark this) then, sure, I will never use the `recycle` option.

**Comment by h908714124** [comment]
Good point @jbgi, seems like `recycle` is less safe than I had hoped. Sorry I missed the `PutBuilderImpl.set(this);` at first. Well...

**Comment by h908714124** [comment]
Maybe switch to `ThreadLocal<Stack<PutBuilderImpl>>`?

**Comment by h908714124** [comment]
@jbgi looks feasible, I will try to fast-track and make a release of it. Finger's crossed.

**Comment by clinuxrulz** [comment]
looks like I'm a bit late in the conversation. This is the code that gave me an issue.

```
        MakeExtensionUtil.getMakeExtensionBayInfosForSide1End1End2SkillionBuilding(
            MakeExtensionUtil.GetMakeExtensionBayInfosForSide1End1End2SkillionBuildingParams
                .with()
                .axes(getAxes(engineeringSystem).toAxes3D())
                .extensionId(Option.some(ID.of(getExtensionId())))
                .length(getLength(engineeringSystem))
                .span(getSpan())
                .height(getHeight())
                .pitch(getPitch())
                .side1Overhang(0.0)
                .end1Overhang(getEnd1Overhang())
                .end2Overhang(getEnd2Overhang())
                .side1BayMarkers(getSideBayMarkers(engineeringSystem).toList())
                .end1BayMarkers(getEnd1BayMarkers(engineeringSystem).toList())
                .end2BayMarkers(getEnd2BayMarkers(engineeringSystem).toList())
        );
```

It turned out ```getSideBayMarkers()``` also used ```GetMakeExtensionBayInfosForSide1End1End2SkillionBuildingParams``` deep down in another building in the scene. It caught me out completely by surprise.

**Comment by h908714124** [comment]
@clinuxrulz I've just released version `1.511`, it should fix the problem. The above rule is now unnecessary; `recycle` should never give incorrect results.

Please note: This uses ref counting under the hood, and `recycle` will fall back to creating new instances if there is more than one ref per thread, like in your "nesting" example. 

Also, the ref counter will be off, if the final step of the builder, or the `done()` method of the updater, is never called. This may happen if one of the step methods throws an exception, for instance. For now, I would recommend not to use `recycle`, unless you're sure that no exceptions will occur at the wrong time.

**Comment by clinuxrulz** [comment]
@h908714124 Cheers.

**Comment by jbgi** [comment]
@h908714124 Here is a variation that should:
- enable reuse in nested call (maximizing reuse I think)
- avoid the situation were the ref count is permanently too high
- track reuse of each builder independently, avoiding unnecessary builder creation.

```java
public final class RequestBuilders {
  private static final ThreadLocal<RequestBuilders> INSTANCE = new ThreadLocal<RequestBuilders>() {
    @Override
    protected RequestBuilders initialValue() {
      return new RequestBuilders();
    }
  };

  private PutBuilderImpl putBuilderImpl = new PutBuilderImpl();

  private RequestBuilders() {
  }

  public static PutBuilder.Path putBuilder() {
    RequestBuilders requestBuilders = INSTANCE.get();
    if (requestBuilders.putBuilderImpl._currently_in_use) {
      requestBuilders.putBuilderImpl = new PutBuilderImpl();
    }
    requestBuilders.putBuilderImpl._currently_in_use = true;
    return requestBuilders.putBuilderImpl;
  }


  private static final class PutBuilderImpl implements PutBuilder.Path, PutBuilder.Body {

    private boolean _currently_in_use;

    private String path;

    PutBuilderImpl() {
    }

    @Override
    public PutBuilder.Body path(String path) {
      this.path = path;
      return this;
    }

    @Override
    public Request body(String body) {
      String path = this.path;
      this.path = null;
      _currently_in_use = false;
      return Request.put(path, body);
    }
  }

  public static final class PutBuilder {
    private PutBuilder() {
      throw new UnsupportedOperationException("no instances");
    }

    public interface Path {
      Body path(String path);
    }

    public interface Body {
      Request body(String body);
    }
  }

}
```

**Comment by h908714124** [comment]
Amazing stuff @jbgi. I'll try to sneak this into the next release.

**Comment by h908714124** [comment]
The improved variation is implemented in version `1.512`. An exception before e.g. `done()` now leads to a new updater instance replacing the previous one, but does not affect subsequent operations.

**Comment by h908714124** [comment]
@jbgi @clinuxrulz FYI I've now implemented this pattern of using a boolean `_currently_in_use` in [auto-builder](https://github.com/h908714124/auto-builder) as well. See the section on "caching" there.
